File: src/deadlock_brain/lineage.py
from __future__ import annotations

import json
import re
import sqlite3
import time
from typing import Any

from deadlock_brain.entity_normalizer import normalize_alias


RENAMED_TO_RE = re.compile(r"^(?:now\s+)?renamed\s+to\s+(?P<new>.+?)(?:[.;,)]|$)", re.IGNORECASE)
RENAMED_X_TO_Y_RE = re.compile(r"^(?:renamed\s+)?(?P<old>.+?)\s+renamed\s+to\s+(?P<new>.+?)(?:[.;,)]|$)", re.IGNORECASE)
RENAMED_OLD_TO_NEW_RE = re.compile(r"^renamed\s+(?P<old>.+?)\s+to\s+(?P<new>.+?)(?:[.;,)]|$)", re.IGNORECASE)
REPLACED_WITH_RE = re.compile(r"^replaced\s+(?P<old>.+?)\s+with\s+(?P<new>.+?)(?:[.;,)]|$)", re.IGNORECASE)
REWORKED_RE = re.compile(r"^(?P<name>.+?)\s+(?:has\s+been\s+)?reworked\b", re.IGNORECASE)

LINEAGE_NAME_LIMIT = 120


def build_entity_lineage(conn: sqlite3.Connection, rebuild: bool = False) -> dict[str, Any]:
    """Build deterministic rename/rework lineage rows from existing patch events."""
    ensure_entity_lineage_table(conn)
    before_total = _count_lineage(conn)
    deleted = 0
    if rebuild:
        cursor = conn.execute("DELETE FROM entity_lineage")
        deleted = int(cursor.rowcount if cursor.rowcount is not None else 0)

    events = conn.execute(
        """
        SELECT id, patch_snapshot_id, patch_external_id, patch_title, patch_url,
               source_kind, posted_at, line_index, section, entity_type,
               entity_name, subject, change_type, normalized_line, raw_line
        FROM patch_events
        WHERE
          change_type IN ('changed', 'rework', 'removed', 'added')
          OR lower(normalized_line) LIKE '%rename%'
          OR lower(normalized_line) LIKE '%rework%'
          OR lower(normalized_line) LIKE '%replaced%'
        ORDER BY id ASC
        """
    ).fetchall()

    processed = 0
    inserted = 0
    by_relation: dict[str, int] = {}
    for row in events:
        event = _row_to_dict(row)
        candidates = extract_lineage_candidates(event)
        processed += 1
        for candidate in candidates:
            if _insert_lineage(conn, candidate):
                inserted += 1
                relation = str(candidate["relation_type"])
                by_relation[relation] = by_relation.get(relation, 0) + 1

    conn.commit()
    return {
        "events_processed": processed,
        "lineage_inserted": inserted,
        "lineage_before": before_total,
        "lineage_total": _count_lineage(conn),
        "deleted_before_build": deleted,
        "by_relation": dict(sorted(by_relation.items())),
        "rebuild": rebuild,
    }


def ensure_entity_lineage_table(conn: sqlite3.Connection) -> None:
    conn.executescript(
        """
        CREATE TABLE IF NOT EXISTS entity_lineage (
          id INTEGER PRIMARY KEY AUTOINCREMENT,
          patch_event_id INTEGER NOT NULL,
          relation_type TEXT NOT NULL,
          source_entity_type TEXT,
          source_name TEXT NOT NULL,
          source_name_norm TEXT NOT NULL,
          target_entity_type TEXT,
          target_name TEXT,
          target_name_norm TEXT,
          owner_entity_type TEXT,
          owner_name TEXT,
          owner_name_norm TEXT,
          confidence REAL NOT NULL,
          metadata_json TEXT NOT NULL DEFAULT '{}',
          created_at INTEGER NOT NULL,
          updated_at INTEGER NOT NULL,
          FOREIGN KEY(patch_event_id) REFERENCES patch_events(id) ON DELETE CASCADE
        );

        CREATE UNIQUE INDEX IF NOT EXISTS idx_entity_lineage_unique
          ON entity_lineage(
            patch_event_id,
            relation_type,
            source_name_norm,
            COALESCE(target_name_norm, ''),
            COALESCE(owner_name_norm, '')
          );

        CREATE INDEX IF NOT EXISTS idx_entity_lineage_source
          ON entity_lineage(source_name_norm);

        CREATE INDEX IF NOT EXISTS idx_entity_lineage_target
          ON entity_lineage(target_name_norm);

        CREATE INDEX IF NOT EXISTS idx_entity_lineage_owner
          ON entity_lineage(owner_name_norm);
        """
    )
    conn.commit()


def extract_lineage_candidates(event: dict[str, Any]) -> list[dict[str, Any]]:
    text = _clean_line(event.get("normalized_line") or event.get("raw_line") or "")
    if not text:
        return []
    event_id = int(event["id"])
    entity_type = _lineage_entity_type(event.get("entity_type"))
    entity_name = _clean_name(event.get("entity_name") or event.get("subject"))
    owner_type = entity_type if entity_type == "hero" else None
    owner_name = entity_name if owner_type else None
    candidates: list[dict[str, Any]] = []

    match = RENAMED_TO_RE.match(text)
    if match and entity_name:
        candidates.append(
            _candidate(
                event_id=event_id,
                relation_type="rename",
                source_type=entity_type,
                source_name=entity_name,
                target_type=entity_type,
                target_name=_clean_name(match.group("new")),
                confidence=0.95,
                event=event,
            )
        )
        return candidates

    for pattern in (RENAMED_X_TO_Y_RE, RENAMED_OLD_TO_NEW_RE):
        match = pattern.match(text)
        if match:
            old_name = _clean_name(match.group("old"))
            new_name = _clean_name(match.group("new"))
            candidates.append(
                _candidate(
                    event_id=event_id,
                    relation_type="rename",
                    source_type=_scoped_child_type(entity_type),
                    source_name=old_name,
                    target_type=_scoped_child_type(entity_type),
                    target_name=new_name,
                    owner_type=owner_type,
                    owner_name=owner_name,
                    confidence=0.9,
                    event=event,
                )
            )
            return candidates

    match = REPLACED_WITH_RE.match(text)
    if match:
        candidates.append(
            _candidate(
                event_id=event_id,
                relation_type="replaced_by",
                source_type=entity_type,
                source_name=_clean_name(match.group("old")),
                target_type=entity_type,
                target_name=_clean_name(match.group("new")),
                confidence=0.75,
                event=event,
            )
        )

    if _is_rework_event(event, text):
        reworked_name = entity_name
        match = REWORKED_RE.match(text)
        if match and not entity_name:
            reworked_name = _clean_name(match.group("name"))
        if reworked_name:
            candidates.append(
                _candidate(
                    event_id=event_id,
                    relation_type="rework",
                    source_type=entity_type,
                    source_name=reworked_name,
                    target_type=entity_type,
                    target_name=reworked_name,
                    owner_type=owner_type if reworked_name != owner_name else None,
                    owner_name=owner_name if reworked_name != owner_name else None,
                    confidence=0.72 if "reworked" in text.casefold() else 0.65,
                    event=event,
                )
            )

    return [candidate for candidate in candidates if _valid_candidate(candidate)]


def related_names_for_query(conn: sqlite3.Connection, query: str, best_match: dict[str, Any] | None = None) -> list[dict[str, Any]]:
    """Return lineage-related names that should be searched together."""
    if not _table_exists(conn, "entity_lineage"):
        return []
    names = {query.strip()}
    if best_match and best_match.get("canonical_name"):
        names.add(str(best_match["canonical_name"]).strip())
    norms = sorted({normalize_alias(name) for name in names if name})
    if not norms:
        return []
    placeholders = ",".join("?" for _ in norms)
    rows = _fetch_all_dicts(
        conn,
        f"""
        SELECT *
        FROM entity_lineage
        WHERE source_name_norm IN ({placeholders})
           OR target_name_norm IN ({placeholders})
           OR owner_name_norm IN ({placeholders})
        ORDER BY confidence DESC, id ASC
        LIMIT ?
        """,
        (*norms, *norms, *norms, LINEAGE_NAME_LIMIT),
    )
    return [_decode_metadata(row) for row in rows]


def lineage_lookup_names(conn: sqlite3.Connection, query: str, best_match: dict[str, Any] | None = None) -> list[str]:
    names = {query.strip()}
    if best_match and best_match.get("canonical_name"):
        names.add(str(best_match["canonical_name"]).strip())
    for row in related_names_for_query(conn, query, best_match):
        for key in ("source_name", "target_name", "owner_name"):
            value = str(row.get(key) or "").strip()
            if value:
                names.add(value)
    return sorted(name for name in names if name)


def _candidate(
    *,
    event_id: int,
    relation_type: str,
    source_type: str | None,
    source_name: str,
    target_type: str | None,
    target_name: str | None,
    confidence: float,
    event: dict[str, Any],
    owner_type: str | None = None,
    owner_name: str | None = None,
) -> dict[str, Any]:
    return {
        "patch_event_id": event_id,
        "relation_type": relation_type,
        "source_entity_type": source_type,
        "source_name": source_name,
        "source_name_norm": normalize_alias(source_name),
        "target_entity_type": target_type,
        "target_name": target_name,
        "target_name_norm": normalize_alias(target_name or "") or None,
        "owner_entity_type": owner_type,
        "owner_name": owner_name,
        "owner_name_norm": normalize_alias(owner_name or "") or None,
        "confidence": confidence,
        "metadata": {
            "patch_title": event.get("patch_title"),
            "patch_url": event.get("patch_url"),
            "posted_at": event.get("posted_at"),
            "source_kind": event.get("source_kind"),
            "line": event.get("normalized_line") or event.get("raw_line"),
        },
    }


def _insert_lineage(conn: sqlite3.Connection, candidate: dict[str, Any]) -> bool:
    now = int(time.time())
    cursor = conn.execute(
        """
        INSERT INTO entity_lineage(
          patch_event_id, relation_type, source_entity_type, source_name, source_name_norm,
          target_entity_type, target_name, target_name_norm, owner_entity_type, owner_name,
          owner_name_norm, confidence, metadata_json, created_at, updated_at
        )
        VALUES(?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)
        ON CONFLICT DO NOTHING
        """,
        (
            candidate["patch_event_id"],
            candidate["relation_type"],
            candidate["source_entity_type"],
            candidate["source_name"],
            candidate["source_name_norm"],
            candidate["target_entity_type"],
            candidate["target_name"],
            candidate["target_name_norm"],
            candidate["owner_entity_type"],
            candidate["owner_name"],
            candidate["owner_name_norm"],
            candidate["confidence"],
            json.dumps(candidate.get("metadata") or {}, ensure_ascii=True, sort_keys=True),
            now,
            now,
        ),
    )
    return cursor.rowcount > 0


def _valid_candidate(candidate: dict[str, Any]) -> bool:
    source = str(candidate.get("source_name") or "").strip()
    target = str(candidate.get("target_name") or "").strip()
    if len(source) < 2 or len(source) > 120:
        return False
    if target and len(target) > 120:
        return False
    return source.casefold() not in {"it", "this", "the item", "the ability"}


def _lineage_entity_type(value: Any) -> str | None:
    text = str(value or "").strip()
    return None if text == "general" else text or None


def _scoped_child_type(parent_type: str | None) -> str | None:
    if parent_type in {"hero", "hero_internal"}:
        return "ability"
    return parent_type


def _is_rework_event(event: dict[str, Any], text: str) -> bool:
    lower = text.casefold()
    return str(event.get("change_type") or "").casefold() == "rework" or "reworked" in lower or "rescaled" in lower


def _clean_line(value: Any) -> str:
    return re.sub(r"\s+", " ", str(value or "").strip())


def _clean_name(value: Any) -> str:
    text = _clean_line(value)
    text = re.sub(r"\s*\([^)]*\)\s*$", "", text).strip()
    return text.strip(" :-\t\"'")


def _count_lineage(conn: sqlite3.Connection) -> int:
    row = conn.execute("SELECT COUNT(*) AS count FROM entity_lineage").fetchone()
    return int(row["count"] if hasattr(row, "keys") else row[0])


def _row_to_dict(row: sqlite3.Row) -> dict[str, Any]:
    return {key: row[key] for key in row.keys()}


def _table_exists(conn: sqlite3.Connection, name: str) -> bool:
    return conn.execute("SELECT 1 FROM sqlite_master WHERE type='table' AND name=?", (name,)).fetchone() is not None


def _fetch_all_dicts(conn: sqlite3.Connection, sql: str, params: tuple[Any, ...] = ()) -> list[dict[str, Any]]:
    cursor = conn.execute(sql, params)
    columns = [column[0] for column in cursor.description or []]
    rows = cursor.fetchall()
    if rows and hasattr(rows[0], "keys"):
        return [{key: row[key] for key in row.keys()} for row in rows]
    return [{key: row[index] for index, key in enumerate(columns)} for row in rows]


def _decode_metadata(row: dict[str, Any]) -> dict[str, Any]:
    metadata = row.pop("metadata_json", "{}")
    try:
        row["metadata"] = json.loads(str(metadata))
    except json.JSONDecodeError:
        row["metadata"] = {}
    return row

File: rust/crates/dbrain-normalize/src/lineage.rs
use std::collections::BTreeMap;

use rusqlite::{params, Connection};
use serde_json::{json, Value};

use crate::util::{clean_subject, normalize_alias, object_json_string, table_count};
use crate::Result;

const LINEAGE_NAME_LIMIT: usize = 120;

#[derive(Debug, Clone, PartialEq)]
pub struct LineageEvent {
    pub id: i64,
    pub patch_title: Option<String>,
    pub patch_url: Option<String>,
    pub source_kind: Option<String>,
    pub posted_at: Option<String>,
    pub entity_type: Option<String>,
    pub entity_name: Option<String>,
    pub subject: Option<String>,
    pub change_type: Option<String>,
    pub normalized_line: Option<String>,
    pub raw_line: Option<String>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct LineageCandidate {
    pub patch_event_id: i64,
    pub relation_type: String,
    pub source_entity_type: Option<String>,
    pub source_name: String,
    pub source_name_norm: String,
    pub target_entity_type: Option<String>,
    pub target_name: Option<String>,
    pub target_name_norm: Option<String>,
    pub owner_entity_type: Option<String>,
    pub owner_name: Option<String>,
    pub owner_name_norm: Option<String>,
    pub confidence: f64,
    pub metadata: BTreeMap<String, Value>,
}

struct CandidateInput {
    relation_type: &'static str,
    source_type: Option<String>,
    source_name: String,
    target_type: Option<String>,
    target_name: Option<String>,
    owner_type: Option<String>,
    owner_name: Option<String>,
    confidence: f64,
}

pub fn enrich_lineage(conn: &Connection, rebuild: bool) -> Result<Value> {
    let before_total = table_count(conn, "entity_lineage")?;
    let deleted = if rebuild {
        conn.execute("DELETE FROM entity_lineage", [])? as i64
    } else {
        0
    };

    let events = load_events(conn)?;
    let mut processed = 0_i64;
    let mut inserted = 0_i64;
    let mut by_relation: BTreeMap<String, i64> = BTreeMap::new();
    for event in events {
        let candidates = extract_lineage_candidates(&event);
        processed += 1;
        for candidate in candidates {
            if insert_lineage(conn, &candidate)? {
                inserted += 1;
                *by_relation.entry(candidate.relation_type.clone()).or_default() += 1;
            }
        }
    }

    Ok(json!({
        "events_processed": processed,
        "lineage_inserted": inserted,
        "lineage_before": before_total,
        "lineage_total": table_count(conn, "entity_lineage")?,
        "deleted_before_build": deleted,
        "by_relation": by_relation,
        "rebuild": rebuild,
    }))
}

pub fn extract_lineage_candidates(event: &LineageEvent) -> Vec<LineageCandidate> {
    let text = clean_line(
        event
            .normalized_line
            .as_deref()
            .or(event.raw_line.as_deref())
            .unwrap_or_default(),
    );
    if text.is_empty() {
        return Vec::new();
    }
    let entity_type = lineage_entity_type(event.entity_type.as_deref());
    let entity_name = clean_name(
        event
            .entity_name
            .as_deref()
            .or(event.subject.as_deref())
            .unwrap_or_default(),
    );
    let entity_name = if entity_name.is_empty() { None } else { Some(entity_name) };
    let owner_type = if entity_type.as_deref() == Some("hero") {
        entity_type.clone()
    } else {
        None
    };
    let owner_name = if owner_type.is_some() { entity_name.clone() } else { None };
    let mut candidates = Vec::new();

    if let Some(new_name) = renamed_to(&text) {
        if let Some(entity_name) = entity_name.clone() {
            candidates.push(candidate(event, CandidateInput {
                relation_type: "rename",
                source_type: entity_type.clone(),
                source_name: entity_name,
                target_type: entity_type.clone(),
                target_name: Some(clean_name(&new_name)),
                owner_type: None,
                owner_name: None,
                confidence: 0.95,
            }));
            return candidates;
        }
    }

    if let Some((old_name, new_name)) = renamed_x_to_y(&text).or_else(|| renamed_old_to_new(&text)) {
        candidates.push(candidate(event, CandidateInput {
            relation_type: "rename",
            source_type: scoped_child_type(entity_type.as_deref()),
            source_name: clean_name(&old_name),
            target_type: scoped_child_type(entity_type.as_deref()),
            target_name: Some(clean_name(&new_name)),
            owner_type: owner_type.clone(),
            owner_name: owner_name.clone(),
            confidence: 0.9,
        }));
        return valid_candidates(candidates);
    }

    if let Some((old_name, new_name)) = replaced_with(&text) {
        candidates.push(candidate(event, CandidateInput {
            relation_type: "replaced_by",
            source_type: entity_type.clone(),
            source_name: clean_name(&old_name),
            target_type: entity_type.clone(),
            target_name: Some(clean_name(&new_name)),
            owner_type: None,
            owner_name: None,
            confidence: 0.75,
        }));
    }

    if is_rework_event(event, &text) {
        let mut reworked_name = entity_name.clone();
        if reworked_name.is_none() {
            reworked_name = reworked_name_from_text(&text);
        }
        if let Some(reworked_name) = reworked_name {
            let keep_owner = owner_name.as_deref() != Some(reworked_name.as_str());
            candidates.push(candidate(event, CandidateInput {
                relation_type: "rework",
                source_type: entity_type.clone(),
                source_name: reworked_name.clone(),
                target_type: entity_type,
                target_name: Some(reworked_name),
                owner_type: if keep_owner { owner_type } else { None },
                owner_name: if keep_owner { owner_name } else { None },
                confidence: if text.to_lowercase().contains("reworked") { 0.72 } else { 0.65 },
            }));
        }
    }

    valid_candidates(candidates)
}

fn load_events(conn: &Connection) -> Result<Vec<LineageEvent>> {
    let mut stmt = conn.prepare(
        r#"
        SELECT id, patch_title, patch_url, source_kind, posted_at, entity_type,
               entity_name, subject, change_type, normalized_line, raw_line
        FROM patch_events
        WHERE
          change_type IN ('changed', 'rework', 'removed', 'added')
          OR lower(normalized_line) LIKE '%rename%'
          OR lower(normalized_line) LIKE '%rework%'
          OR lower(normalized_line) LIKE '%replaced%'
        ORDER BY id ASC
        "#,
    )?;
    let rows = stmt.query_map([], |row| {
        Ok(LineageEvent {
            id: row.get("id")?,
            patch_title: row.get("patch_title")?,
            patch_url: row.get("patch_url")?,
            source_kind: row.get("source_kind")?,
            posted_at: row.get("posted_at")?,
            entity_type: row.get("entity_type")?,
            entity_name: row.get("entity_name")?,
            subject: row.get("subject")?,
            change_type: row.get("change_type")?,
            normalized_line: row.get("normalized_line")?,
            raw_line: row.get("raw_line")?,
        })
    })?;
    let mut out = Vec::new();
    for row in rows {
        out.push(row?);
    }
    Ok(out)
}

fn candidate(event: &LineageEvent, input: CandidateInput) -> LineageCandidate {
    let mut metadata = BTreeMap::new();
    metadata.insert("patch_title".to_string(), option_json(event.patch_title.clone()));
    metadata.insert("patch_url".to_string(), option_json(event.patch_url.clone()));
    metadata.insert("posted_at".to_string(), option_json(event.posted_at.clone()));
    metadata.insert("source_kind".to_string(), option_json(event.source_kind.clone()));
    metadata.insert(
        "line".to_string(),
        option_json(event.normalized_line.clone().or_else(|| event.raw_line.clone())),
    );
    let target_name_norm = input
        .target_name
        .as_deref()
        .map(normalize_alias)
        .filter(|value| !value.is_empty());
    let owner_name_norm = input
        .owner_name
        .as_deref()
        .map(normalize_alias)
        .filter(|value| !value.is_empty());
    LineageCandidate {
        patch_event_id: event.id,
        relation_type: input.relation_type.to_string(),
        source_entity_type: input.source_type,
        source_name_norm: normalize_alias(&input.source_name),
        source_name: input.source_name,
        target_entity_type: input.target_type,
        target_name: input.target_name,
        target_name_norm,
        owner_entity_type: input.owner_type,
        owner_name: input.owner_name,
        owner_name_norm,
        confidence: input.confidence,
        metadata,
    }
}

fn insert_lineage(conn: &Connection, candidate: &LineageCandidate) -> Result<bool> {
    let now = crate::util::now()?;
    let metadata_json = object_json_string(candidate.metadata.clone())?;
    let changed = conn.execute(
        r#"
        INSERT INTO entity_lineage(
          patch_event_id, relation_type, source_entity_type, source_name, source_name_norm,
          target_entity_type, target_name, target_name_norm, owner_entity_type, owner_name,
          owner_name_norm, confidence, metadata_json, created_at, updated_at
        )
        VALUES(?1,?2,?3,?4,?5,?6,?7,?8,?9,?10,?11,?12,?13,?14,?15)
        ON CONFLICT DO NOTHING
        "#,
        params![
            candidate.patch_event_id,
            candidate.relation_type,
            candidate.source_entity_type,
            candidate.source_name,
            candidate.source_name_norm,
            candidate.target_entity_type,
            candidate.target_name,
            candidate.target_name_norm,
            candidate.owner_entity_type,
            candidate.owner_name,
            candidate.owner_name_norm,
            candidate.confidence,
            metadata_json,
            now,
            now,
        ],
    )?;
    Ok(changed > 0)
}

fn valid_candidates(candidates: Vec<LineageCandidate>) -> Vec<LineageCandidate> {
    candidates.into_iter().filter(valid_candidate).collect()
}

fn valid_candidate(candidate: &LineageCandidate) -> bool {
    let source = candidate.source_name.trim();
    let target = candidate.target_name.as_deref().unwrap_or_default().trim();
    if source.len() < 2 || source.len() > LINEAGE_NAME_LIMIT {
        return false;
    }
    if !target.is_empty() && target.len() > LINEAGE_NAME_LIMIT {
        return false;
    }
    !matches!(source.to_lowercase().as_str(), "it" | "this" | "the item" | "the ability")
}

fn lineage_entity_type(value: Option<&str>) -> Option<String> {
    let text = value.unwrap_or_default().trim();
    if text.is_empty() || text == "general" {
        None
    } else {
        Some(text.to_string())
    }
}

fn scoped_child_type(parent_type: Option<&str>) -> Option<String> {
    if matches!(parent_type, Some("hero" | "hero_internal")) {
        Some("ability".to_string())
    } else {
        parent_type.map(ToString::to_string)
    }
}

fn is_rework_event(event: &LineageEvent, text: &str) -> bool {
    let lower = text.to_lowercase();
    event
        .change_type
        .as_deref()
        .map(|value| value.eq_ignore_ascii_case("rework"))
        .unwrap_or(false)
        || lower.contains("reworked")
        || lower.contains("rescaled")
}

fn clean_line(value: &str) -> String {
    value.split_whitespace().collect::<Vec<_>>().join(" ")
}

fn clean_name(value: &str) -> String {
    let mut text = clean_line(value);
    if text.ends_with(')') {
        if let Some(index) = text.rfind('(') {
            text = text[..index].trim().to_string();
        }
    }
    clean_subject(Some(text.trim_matches(|ch: char| {
        ch == ' ' || ch == ':' || ch == '-' || ch == '\t' || ch == '"' || ch == '\''
    })))
}

fn renamed_to(text: &str) -> Option<String> {
    let lower = text.to_lowercase();
    let prefix = if lower.starts_with("now renamed to ") {
        "now renamed to "
    } else if lower.starts_with("renamed to ") {
        "renamed to "
    } else {
        return None;
    };
    Some(sentence_tail(&text[prefix.len()..]))
}

fn renamed_x_to_y(text: &str) -> Option<(String, String)> {
    let lower = text.to_lowercase();
    let marker = " renamed to ";
    if lower.starts_with("renamed to ") {
        return None;
    }
    let index = lower.find(marker)?;
    let old_name = text[..index].trim();
    let new_name = &text[index + marker.len()..];
    if old_name.is_empty() {
        None
    } else {
        Some((old_name.to_string(), sentence_tail(new_name)))
    }
}

fn renamed_old_to_new(text: &str) -> Option<(String, String)> {
    let lower = text.to_lowercase();
    let rest = lower.strip_prefix("renamed ")?;
    let index = rest.find(" to ")?;
    let old_start = "renamed ".len();
    let old_end = old_start + index;
    let new_start = old_end + " to ".len();
    Some((text[old_start..old_end].trim().to_string(), sentence_tail(&text[new_start..])))
}

fn replaced_with(text: &str) -> Option<(String, String)> {
    let lower = text.to_lowercase();
    let rest = lower.strip_prefix("replaced ")?;
    let index = rest.find(" with ")?;
    let old_start = "replaced ".len();
    let old_end = old_start + index;
    let new_start = old_end + " with ".len();
    Some((text[old_start..old_end].trim().to_string(), sentence_tail(&text[new_start..])))
}

fn reworked_name_from_text(text: &str) -> Option<String> {
    let lower = text.to_lowercase();
    if let Some(index) = lower.find(" has been reworked") {
        let name = clean_name(&text[..index]);
        return (!name.is_empty()).then_some(name);
    }
    if let Some(index) = lower.find(" reworked") {
        let name = clean_name(&text[..index]);
        return (!name.is_empty()).then_some(name);
    }
    None
}

fn sentence_tail(value: &str) -> String {
    value
        .trim()
        .trim_end_matches(['.', ';', ',', ')'])
        .trim()
        .to_string()
}

fn option_json(value: Option<String>) -> Value {
    value.map_or(Value::Null, Value::String)
}
